File: Hisab/Services/Queries.swift
import Foundation
import SwiftData
import HisabCore

/// Fetch helpers bridging SwiftData rows to HisabCore value types.
@MainActor
enum Queries {
    static func documents(_ ctx: ModelContext) -> [DocumentSummary] {
        let docs = (try? ctx.fetch(FetchDescriptor<StoredDocument>())) ?? []
        return docs.map { DocumentSummary(id: $0.uuid, source: $0.source, period: $0.period) }
    }

    static func allTransactions(_ ctx: ModelContext) -> [StoredTransaction] {
        (try? ctx.fetch(FetchDescriptor<StoredTransaction>(sortBy: [SortDescriptor(\.date, order: .reverse)]))) ?? []
    }

    /// Seeds the rule table from Categorizer defaults on first launch.
    static func categoryRules(_ ctx: ModelContext) -> [CategoryRule] {
        var stored = (try? ctx.fetch(FetchDescriptor<StoredCategoryRule>(sortBy: [SortDescriptor(\.sortOrder)]))) ?? []
        if stored.isEmpty {
            for (index, rule) in Categorizer.seedRules.enumerated() {
                ctx.insert(StoredCategoryRule(pattern: rule.pattern, category: rule.category, sortOrder: index))
            }
            try? ctx.save()
            stored = (try? ctx.fetch(FetchDescriptor<StoredCategoryRule>(sortBy: [SortDescriptor(\.sortOrder)]))) ?? []
        }
        return stored.map(\.asRule)
    }

    static func category(of txn: StoredTransaction, rules: [CategoryRule]) -> String {
        txn.categoryOverride ?? Categorizer.category(for: "\(txn.counterparty) \(txn.narration)", rules: rules)
    }

    // MARK: pure projections over already-fetched rows
    // Views feed these from @Query so SwiftData changes invalidate them automatically —
    // a view that computes from a ModelContext fetch alone never refreshes (the bug that
    // left the Transactions tab stale on-device).

    static func rules(from rows: [StoredCategoryRule]) -> [CategoryRule] {
        rows.map(\.asRule)
    }

    /// UUIDs of bank rows confirmed as the bank-side copy of an app payment.
    static func matchedBankUUIDs(in matches: [StoredMatch]) -> Set<UUID> {
        Set(matches.map(\.bankUUID))
    }

    /// UUIDs of cross-bank self-transfer pairs (HDFC↔IDFC internal movements).
    static func selfTransferUUIDs(in txns: [StoredTransaction]) -> Set<UUID> {
        let bank = txns.filter { $0.source.kind == .bank }
        return SelfTransfers.detect(bank: bank.map {
            (ReconTxn(id: $0.uuid, date: $0.date, amountPaise: $0.amountPaise,
                      direction: $0.direction, reference: $0.reference), $0.source)
        })
    }

    /// The recorded history: every payment-app transaction, plus bank rows that have no
    /// app counterpart. Matched bank rows are reconciliation evidence, not records.
    static func visible(_ txns: [StoredTransaction], matches: [StoredMatch]) -> [StoredTransaction] {
        let matched = matchedBankUUIDs(in: matches)
        return txns.filter { !matched.contains($0.uuid) }
    }

    /// Counted analytics rows: visible history minus self transfers.
    static func analytics(txns: [StoredTransaction], matches: [StoredMatch],
                          rules: [CategoryRule]) -> [AnalyticsTxn] {
        let selfTransfers = selfTransferUUIDs(in: txns)
        return visible(txns, matches: matches)
            .filter { !selfTransfers.contains($0.uuid) }
            .map { txn in
                AnalyticsTxn(month: txn.month,
                             amountPaise: txn.amountPaise,
                             direction: txn.direction,
                             category: effectiveCategory(of: txn, rules: rules, selfTransfers: []),
                             merchant: txn.counterparty,
                             sourceKind: txn.source.kind)
            }
    }

    static func grid(documents: [StoredDocument], pinned: [PinnedMonth]) -> CoverageGrid {
        CoverageGrid.derive(
            documents: documents.map { DocumentSummary(id: $0.uuid, source: $0.source, period: $0.period) },
            pinnedMonths: Set(pinned.map(\.yearMonth)))
    }

    static func reconProjection(_ txns: [StoredTransaction],
                                month: YearMonth) -> (app: [ReconTxn], bank: [ReconTxn]) {
        let monthTxns = txns.filter { $0.month == month }
        func projected(_ kind: SourceKind) -> [ReconTxn] {
            monthTxns.filter { $0.source.kind == kind }
                .map { ReconTxn(id: $0.uuid, date: $0.date, amountPaise: $0.amountPaise,
                                direction: $0.direction, reference: $0.reference) }
        }
        return (projected(.paymentApp), projected(.bank))
    }

    // MARK: context-based variants (import pipeline and one-shot sheets)

    /// Display/analytics category. Bank-only rows fall back to Miscellaneous rather
    /// than Uncategorized; self transfers are labeled as such.
    static func effectiveCategory(of txn: StoredTransaction, rules: [CategoryRule],
                                  selfTransfers: Set<UUID>) -> String {
        if selfTransfers.contains(txn.uuid) { return Categorizer.selfTransfer }
        if let override = txn.categoryOverride { return override }
        let auto = Categorizer.category(for: "\(txn.counterparty) \(txn.narration)", rules: rules)
        if auto == Categorizer.uncategorized && txn.source.kind == .bank {
            return Categorizer.miscellaneous
        }
        return auto
    }

    static func reconTxns(_ ctx: ModelContext, month: YearMonth) -> (app: [ReconTxn], bank: [ReconTxn]) {
        reconProjection(allTransactions(ctx), month: month)
    }

    static func matches(_ ctx: ModelContext, month: YearMonth) -> [StoredMatch] {
        let key = month.description
        let descriptor = FetchDescriptor<StoredMatch>(predicate: #Predicate { $0.monthKey == key })
        return (try? ctx.fetch(descriptor)) ?? []
    }

    static func pinned(_ ctx: ModelContext) -> Set<YearMonth> {
        let rows = (try? ctx.fetch(FetchDescriptor<PinnedMonth>())) ?? []
        return Set(rows.map(\.yearMonth))
    }

    static func coverageGrid(_ ctx: ModelContext) -> CoverageGrid {
        CoverageGrid.derive(documents: documents(ctx), pinnedMonths: pinned(ctx))
    }

    /// Replaces the stored reconciliation for one month with a fresh run.
    /// Self-transfer rows are internal movements — kept out of the bank side.
    static func recomputeMatches(_ ctx: ModelContext, month: YearMonth) {
        for row in matches(ctx, month: month) { ctx.delete(row) }
        let (app, rawBank) = reconTxns(ctx, month: month)
        let selfTransfers = selfTransferUUIDs(in: allTransactions(ctx))
        let bank = rawBank.filter { !selfTransfers.contains($0.id) }
        guard !app.isEmpty, !bank.isEmpty else { return }
        let result = Reconciler.reconcile(app: app, bank: bank)
        for pair in result.matches {
            ctx.insert(StoredMatch(appUUID: pair.appID, bankUUID: pair.bankID, tier: pair.tier, month: month))
        }
    }
}

File: hisab_flutter/lib/services/queries.dart
/// Projections and DB helpers, mirroring Hisab/Services/Queries.swift.
/// Pure list-based projections feed the widgets; DB variants serve the
/// import pipeline.
library;

import 'dart:math';

import 'package:hisab_core/hisab_core.dart';

import '../storage/database.dart';

String newId() {
  final random = Random.secure();
  final hex = List.generate(16, (_) => random.nextInt(16).toRadixString(16));
  return '${DateTime.now().microsecondsSinceEpoch.toRadixString(16)}-${hex.join()}';
}

Direction directionOf(StoredTransaction txn) =>
    txn.direction == 'credit' ? Direction.credit : Direction.debit;

Source sourceOf(StoredTransaction txn) => Source(txn.sourceRaw);

DateTime dateOf(StoredTransaction txn) =>
    DateTime.fromMillisecondsSinceEpoch(txn.dateMs, isUtc: true);

class Queries {
  static List<CategoryRule> rules(List<StoredCategoryRule> rows) {
    final sorted = List.of(rows)
      ..sort((a, b) => a.sortOrder.compareTo(b.sortOrder));
    return [
      for (final row in sorted)
        CategoryRule(id: row.id, pattern: row.pattern, category: row.category)
    ];
  }

  static Set<String> matchedBankUuids(List<StoredMatche> matches) =>
      matches.map((m) => m.bankUuid).toSet();

  static Set<String> selfTransferUuids(List<StoredTransaction> txns) {
    final bank = [
      for (final txn in txns)
        if (sourceOf(txn).kind == SourceKind.bank)
          BankTxn(
              ReconTxn(
                  id: txn.uuid,
                  date: dateOf(txn),
                  amountPaise: txn.amountPaise,
                  direction: directionOf(txn),
                  reference: txn.reference),
              sourceOf(txn))
    ];
    return SelfTransfers.detect(bank);
  }

  /// Matched bank rows are reconciliation evidence — hidden from history.
  static List<StoredTransaction> visible(
      List<StoredTransaction> txns, List<StoredMatche> matches) {
    final matched = matchedBankUuids(matches);
    return [
      for (final txn in txns)
        if (sourceOf(txn).kind != SourceKind.bank ||
            !matched.contains(txn.uuid))
          txn
    ];
  }

  /// Display/analytics category. Bank-only rows fall back to Miscellaneous;
  /// self transfers are labeled as such.
  static String effectiveCategory(StoredTransaction txn,
      List<CategoryRule> ruleList, Set<String> selfTransfers) {
    if (selfTransfers.contains(txn.uuid)) return Categorizer.selfTransfer;
    final override = txn.categoryOverride;
    if (override != null) return override;
    final auto = Categorizer.category(
        '${txn.counterparty} ${txn.narration}', ruleList);
    if (auto == Categorizer.uncategorized &&
        sourceOf(txn).kind == SourceKind.bank) {
      return Categorizer.miscellaneous;
    }
    return auto;
  }

  static List<AnalyticsTxn> analytics(List<StoredTransaction> txns,
      List<StoredMatche> matches, List<CategoryRule> ruleList) {
    final selfTransfers = selfTransferUuids(txns);
    return [
      for (final txn in visible(txns, matches))
        if (!selfTransfers.contains(txn.uuid))
          AnalyticsTxn(
            month: YearMonth.fromDate(dateOf(txn)),
            amountPaise: txn.amountPaise,
            direction: directionOf(txn),
            category: effectiveCategory(txn, ruleList, selfTransfers),
            merchant: txn.counterparty,
            sourceKind: sourceOf(txn).kind,
          )
    ];
  }

  static CoverageGrid grid(
      List<StoredDocument> documents, List<PinnedMonth> pins) {
    return CoverageGrid.derive(
      documents: [
        for (final doc in documents)
          DocumentSummary(
              id: doc.id,
              source: Source(doc.sourceRaw),
              period: DatePeriod(
                  DateTime.fromMillisecondsSinceEpoch(doc.periodStartMs,
                      isUtc: true),
                  DateTime.fromMillisecondsSinceEpoch(doc.periodEndMs,
                      isUtc: true)))
      ],
      pinnedMonths: {
        for (final pin in pins) _month(pin.monthKey)
      },
    );
  }

  static YearMonth _month(String key) {
    final parts = key.split('-');
    return YearMonth(int.parse(parts[0]), int.parse(parts[1]));
  }

  static (List<ReconTxn>, List<ReconTxn>) reconProjection(
      List<StoredTransaction> txns, YearMonth month) {
    List<ReconTxn> projected(SourceKind kind) => [
          for (final txn in txns)
            if (sourceOf(txn).kind == kind &&
                YearMonth.fromDate(dateOf(txn)) == month)
              ReconTxn(
                  id: txn.uuid,
                  date: dateOf(txn),
                  amountPaise: txn.amountPaise,
                  direction: directionOf(txn),
                  reference: txn.reference)
        ];
    return (projected(SourceKind.paymentApp), projected(SourceKind.bank));
  }

  /// Projection feeding SuggestionEngine.
  static List<SpendRecord> suggestionRecords(List<StoredTransaction> txns,
      List<StoredMatche> matches, List<CategoryRule> ruleList) {
    final selfTransfers = selfTransferUuids(txns);
    return [
      for (final txn in visible(txns, matches))
        SpendRecord(
          merchant: txn.counterparty.isEmpty ? txn.narration : txn.counterparty,
          amountPaise: txn.amountPaise,
          date: dateOf(txn),
          direction: directionOf(txn),
          effectiveCategory: effectiveCategory(txn, ruleList, selfTransfers),
        )
    ];
  }

  // ---- DB variants

  /// Additive seeding: any ruleset pattern the user doesn't already have is
  /// inserted; existing rules — including edited ones — are never touched.
  static Future<List<CategoryRule>> categoryRules(
      AppDatabase db, Ruleset ruleset) async {
    var stored = await db.select(db.storedCategoryRules).get();
    final existing = {for (final r in stored) r.pattern.toLowerCase()};
    final missing = [
      for (final rule in ruleset.rules)
        if (!existing.contains(rule.pattern.toLowerCase())) rule
    ];
    if (missing.isNotEmpty) {
      var order = stored.isEmpty
          ? 0
          : stored.map((r) => r.sortOrder).reduce(max) + 1;
      await db.batch((batch) {
        for (final rule in missing) {
          batch.insert(
              db.storedCategoryRules,
              StoredCategoryRulesCompanion.insert(
                  id: newId(),
                  pattern: rule.pattern,
                  category: rule.category,
                  sortOrder: order++));
        }
      });
      stored = await db.select(db.storedCategoryRules).get();
    }
    return rules(stored);
  }

  static Future<void> recomputeMatches(AppDatabase db, YearMonth month) async {
    final txns = await db.select(db.storedTransactions).get();
    final (app, bank) = reconProjection(txns, month);
    final result = Reconciler.reconcile(app: app, bank: bank);
    final key = month.toString();
    await (db.delete(db.storedMatches)
          ..where((m) => m.monthKey.equals(key)))
        .go();
    await db.batch((batch) {
      for (final match in result.matches) {
        batch.insert(
            db.storedMatches,
            StoredMatchesCompanion.insert(
                id: newId(),
                monthKey: key,
                appUuid: match.appID,
                bankUuid: match.bankID,
                tier: match.tier.name));
      }
    });
  }
}
